src/dbus_api/blockdev.rs

```diff
@@ -0,0 +1,285 @@
+/// found by means of the getter method which takes a reference to a
+/// blockdev and obtains the property from the blockdev.
+fn get_blockdev_property<F>(i: &mut IterAppend,
+                            p: &PropInfo<MTFn<TData>, TData>,
+                            getter: F)
+                            -> Result<(), MethodErr>
+    where F: Fn(&BlockDev) -> Result<MessageItem, MethodErr>
+{
+    let dbus_context = p.tree.get_data();
+    let object_path = p.path.get_name();
+
+    let blockdev_path = p.tree
+        .get(object_path)
+        .expect("tree must contain implicit argument");
+
+    let blockdev_data =
+        blockdev_path
+            .get_data()
+            .as_ref()
+            .ok_or_else(|| MethodErr::failed(&format!("no data for object path {}", object_path)))?;
+
+    let pool_path = p.tree
+        .get(&blockdev_data.parent)
+        .ok_or_else(|| {
+                        MethodErr::failed(&format!("no path for parent object path {}",
+                                                   &blockdev_data.parent))
+                    })?;
+
+    let pool_uuid = pool_path
+        .get_data()
+        .as_ref()
+        .ok_or_else(|| MethodErr::failed(&format!("no data for object path {}", object_path)))?
+        .uuid;
+
+    let engine = dbus_context.engine.borrow();
+    let pool =
+        engine
+            .get_pool(&pool_uuid)
+            .ok_or_else(|| {
+                            MethodErr::failed(&format!("no pool corresponding to uuid {}",
+                                                       &pool_uuid))
+                        })?;
+    let blockdev_uuid = &blockdev_data.uuid;
+    let blockdev =
+        pool.get_blockdev(blockdev_uuid)
+            .ok_or_else(|| MethodErr::failed(&format!("noblockdev with uuid {}", &blockdev_uuid)))?;
+    i.append(getter(blockdev)?);
+    Ok(())
+}
+
+/// Get the devnode for an object path.
+fn get_blockdev_devnode(i: &mut IterAppend,
+                        p: &PropInfo<MTFn<TData>, TData>)
+                        -> Result<(), MethodErr> {
+    get_blockdev_property(i,
+                          p,
+                          |p| Ok(MessageItem::Str(format!("{}", p.devnode().display()))))
+}
+
+fn get_blockdev_hardware_id(i: &mut IterAppend,
+                            p: &PropInfo<MTFn<TData>, TData>)
+                            -> Result<(), MethodErr> {
+    fn get_hardware_id(blockdev: &BlockDev) -> Result<MessageItem, MethodErr> {
+        match *blockdev.hardware_id() {
+            Some(ref id) => Ok(MessageItem::Str(id.to_owned())),
+            None => Ok(MessageItem::Str("".to_owned())),
+        }
+    }
+
+    get_blockdev_property(i, p, get_hardware_id)
+}
+
+fn get_blockdev_user_id(i: &mut IterAppend,
+                        p: &PropInfo<MTFn<TData>, TData>)
+                        -> Result<(), MethodErr> {
+    fn get_user_id(blockdev: &BlockDev) -> Result<MessageItem, MethodErr> {
+        match *blockdev.user_id() {
```

Comment:
You could be so much more rustic here, something like:

```Ok(MessageItem::Str(*blockdev.user_id().or("").map(|x| x.to_owned())))```

Reply:
cool. This also led to changing `user/hardware_info()` to return `Option<&str>` instead of `&Option<String>`, which is more convenient all around.